src/lib/str/ares__buf.c

```diff
@@ -884,6 +892,61 @@ ares_status_t ares__buf_split(ares__buf_t *buf, const unsigned char *delims,
   return status;
 }
 
+ares_status_t ares__buf_split_str(ares__buf_t *buf, const unsigned char *delims,
+                                  size_t delims_len, ares__buf_split_t flags,
+                                  size_t max_sections, char ***strs,
+                                  size_t *nstrs)
+{
+  ares_status_t       status;
+  ares__llist_t      *list = NULL;
+  ares__llist_node_t *node;
+  ares__array_t      *arr  = NULL;
+
+  if (strs == NULL || nstrs == NULL) {
+    return ARES_EFORMERR;
+  }
+
+  *strs  = NULL;
+  *nstrs = 0;
+
+  status = ares__buf_split(buf, delims, delims_len, flags, max_sections, &list);
+  if (status != ARES_SUCCESS) {
+    goto done;
+  }
+
+  arr = ares__array_create(sizeof(char *), ares_free);
+  if (arr == NULL) {
+    status = ARES_ENOMEM;
+    goto done;
+  }
+
+  for (node = ares__llist_node_first(list); node != NULL;
+       ares__llist_node_next(node)) {
+    ares__buf_t *lbuf = ares__llist_node_val(node);
+    char        *str  = NULL;
+
+    status = ares__buf_fetch_str_dup(lbuf, ares__buf_len(lbuf), &str);
+    if (status != ARES_SUCCESS) {
+      goto done;
+    }
+
+    status = ares__array_insertdata_last(arr, &str);
+    if (status != ARES_SUCCESS) {
+      ares_free(str);
+      goto done;
+    }
+  }
+
+done:
+  ares__llist_destroy(list);
+  if (status == ARES_SUCCESS) {
+    *strs = ares__array_finish(arr, nstrs);
+  } else {
+    ares__array_destroy(arr);
```

Comment:
I'm not sure I'm missing something in my calling code, but trying to use `ares__buf_split_str()` like in the example at https://github.com/c-ares/c-ares/pull/856#issuecomment-2308956347 I get a `free(): double free detected in tcache 2` at this stage.

Reply:
I need to add some tests still before I merge this.  Infact, this hasn't been tested at all yet, so its not unlikely there's some issues hiding out in this PR.

Reply:
Ah alright, I'll stay tuned then. :)

Reply:
ok, fixed now :)